### ISSUE TITLE
Issues with Websocket layer stability. Times out in 5 to 60 seconds based on topic traffic

### ISSUE THREAD

**Issue by lotros** [emqx/emqx#13943]
### What happened?

After updating to EMQX 5.8, we've started to get unusual behavior from our websocket MQTT connection to the broker.

Basically, if we try to listen to heavly used topic, our MQTT connection through websocket keeps timing out and reconnecting.
The larger the throughput of the topic, the faster it times us out.

If we try to connect to the EMQX through the TCP layer, the connection remains stable regardless of the throughput of the topic we are listening to.

We saw similiar issues to ours in other comments, but most of them were connected to the broker timing out, but in this case the broker remains stable, its only the websocket MQTT clients that time out.

### What did you expect to happen?

To have stable communication through websocket MQTT layer.

### How can we reproduce it (as minimally and precisely as possible)?

Easiest approach is to have a 1 or few clients generate about 2000-2500 messages per second. Once that is done, you need to connect with Websocket based MQTT client to some topic that receives about 2 or 3 messages per second. Depending on the messages per second you should loose connection in 5 to 60 seconds.

### EMQX version

sysdescr  : EMQX
version   : 5.8.0
datetime  : 2024-10-07T15:26:25.032550352+03:00
uptime    : 4 days, 16 hours, 4 minutes, 46 seconds

### OS version

PRETTY_NAME="Ubuntu 24.04.1 LTS"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04.1 LTS (Noble Numbat)"
VERSION_CODENAME=noble
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=noble
LOGO=ubuntu-logo
Linux icon-main-server 6.8.0-45-generic #45-Ubuntu SMP PREEMPT_DYNAMIC Fri Aug 30 12:02:04 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Log files

<details>

</details>

**Comment by thalesmg** [comment]
Hi @lotros .

I'm trying to reproduce the behavior, but haven't managed so far.  Could you provide a few more clarifications?

> Basically, if we try to listen to heavly used topic, our MQTT connection through websocket keeps timing out and reconnecting.
> ... you need to connect with Websocket based MQTT client to some topic that receives about 2 or 3 messages per second. 

Do I understand correctly that the WS clients don't actually subscribe to heavily published topics?  Could you provide more precise example of topics and topic filters used when publishing and subscribing, respectively?

Also:
- Do you have any logs from EMQX?
- Which WebSocket MQTT are you using to test?

If the WS MQTT client cannot cope with the MQTT message throughput and send their periodic `PING` messages, it could be one reason for the disconnections.

**Comment by thalesmg** [comment]
Also, logs could help identify if the issue is this known bug that was [recently fixed](https://github.com/emqx/emqx/pull/13754) related to WS connections.  If it's the same, the fix should land in EMQX 5.8.1.

**Comment by thalesmg** [comment]
The fact that I'm not able to reproduce would corroborate this hypothesis: I've tried to use the current `master`, which contains the fix.  :sweat_smile:

**Comment by lotros** [comment]
The timing out happens if you subscribe to a topic with moderate/high throughput.

Mainly our platform uses MQTT.js in an Android app. But to be sure, we tested with multiple MQTT clients using websocket connections and the results are the same on every client we attempt to use.

I made a quick video, showing the regular disconnecting on a fairly medium output topic:
https://www.youtube.com/watch?v=z5kcCY3C0FY

Here are the logs, that are shown each time there is disconnect from the client side:
Oct 07 21:02:55 icon-main-server bash[55407]: 2024-10-07T21:02:55.101681+03:00 [error] supervisor: {esockd_connection_sup,<0.1291275.0>}, errorContext: connection_shutdown, reason: #{cause => malformed_packet,header_type => 4}, offender: [{pid,<0.1291275.0>},{name,connection},{mfargs,{emqx_connection,start_link,[#{listener => {tcp,default},limiter => #{connection => #{initial => 0,rate => infinity,burst => 0}},enable_authn => true,zone => default}]}}]

Oct 07 21:02:55 icon-main-server bash[55407]: 2024-10-07T21:02:55.849707+03:00 [error] crasher: initial call: cowboy_tls:connection_process/4, pid: <0.1293847.0>, registered_name: [], error: {{badmatch,9638},[{emqx_ws_connection,handle_outgoing,2,[{file,"emqx_ws_connection.erl"},{line,814}]},{emqx_ws_connection,websocket_info,2,[{file,"emqx_ws_connection.erl"},{line,998}]},{cowboy_websocket,handler_call,6,[{file,"cowboy_websocket.erl"},{line,487}]},{proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,251}]}]}, ancestors: [<0.2987.0>,<0.2986.0>,ranch_sup,<0.2513.0>], message_queue_len: 0, messages: [], links: [<0.2987.0>], dictionary: [{send_msg,201},{send_cnt,203},{recv_oct,76},{outgoing_pubs,0},{{#{action_type => subscribe,qos => 0},<<"devices/#">>},{allow,1728324175380}},{recv_cnt,2},{'$logger_metadata$',#{peername => "46.10.251.197:40417",username => <<"tempex">>,clientid => <<"mqttx_1c8ea6e4">>}},{outgoing_bytes,0},{authz_keys_q,{[{#{action_type => subscribe,qos => 0},<<"devices/#">>}],[]}},{send_oct,9638},{send_pkt,203},{'send_msg.qos0',201},{authz_cache_size,1},{incoming_bytes,76},{recv_pkt,2},{guid,{1728324175324641,268577191214615,0}}], trap_exit: true, status: running, heap_size: 1598, stack_size: 28, reductions: 117028; neighbours:

Oct 07 21:02:55 icon-main-server bash[55407]: 2024-10-07T21:02:55.850570+03:00 [error] Ranch listener 'wss:default' had connection process started with cowboy_tls:start_link/4 at <0.1293847.0> exit with reason: {{badmatch,9638},[{emqx_ws_connection,handle_outgoing,2,[{file,"emqx_ws_connection.erl"},{line,814}]},{emqx_ws_connection,websocket_info,2,[{file,"emqx_ws_connection.erl"},{line,998}]},{cowboy_websocket,handler_call,6,[{file,"cowboy_websocket.erl"},{line,487}]},{proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,251}]}]}

Oct 07 21:03:00 icon-main-server bash[55407]: 2024-10-07T21:03:00.623776+03:00 [error] crasher: initial call: cowboy_tls:connection_process/4, pid: <0.1293883.0>, registered_name: [], error: {{badmatch,9515},[{emqx_ws_connection,handle_outgoing,2,[{file,"emqx_ws_connection.erl"},{line,814}]},{emqx_ws_connection,websocket_info,2,[{file,"emqx_ws_connection.erl"},{line,998}]},{cowboy_websocket,handler_call,6,[{file,"cowboy_websocket.erl"},{line,487}]},{proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,251}]}]}, ancestors: [<0.2987.0>,<0.2986.0>,ranch_sup,<0.2513.0>], message_queue_len: 0, messages: [], links: [<0.2987.0>], dictionary: [{send_msg,201},{send_cnt,203},{recv_oct,76},{outgoing_pubs,0},{{#{action_type => subscribe,qos => 0},<<"devices/#">>},{allow,1728324180169}},{recv_cnt,2},{'$logger_metadata$',#{peername => "46.10.251.197:40657",username => <<"tempex">>,clientid => <<"mqttx_1c8ea6e4">>}},{outgoing_bytes,0},{authz_keys_q,{[{#{action_type => subscribe,qos => 0},<<"devices/#">>}],[]}},{send_oct,9515},{send_pkt,203},{'send_msg.qos0',201},{authz_cache_size,1},{incoming_bytes,76},{recv_pkt,2},{guid,{1728324180114491,268577191214651,0}}], trap_exit: true, status: running, heap_size: 2586, stack_size: 28, reductions: 115054; neighbours:

Oct 07 21:03:00 icon-main-server bash[55407]: 2024-10-07T21:03:00.624780+03:00 [error] 
Ranch listener 'wss:default' had connection process started with cowboy_tls:start_link/4 at <0.1293883.0> exit with reason: {{badmatch,9515},[{emqx_ws_connection,handle_outgoing,2,[{file,"emqx_ws_connection.erl"},{line,814}]},{emqx_ws_connection,websocket_info,2,[{file,"emqx_ws_connection.erl"},{line,998}]},{cowboy_websocket,handler_call,6,[{file,"cowboy_websocket.erl"},{line,487}]},{proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,251}]}]}

If the problem is the one you mentioned, I know its a bit rude to ask, but what is the timeframe of the release of 5.8.1 ? Since we are using the EMQX in a production level project and its causing some issues to end users ? If I can provide any more information, just tell me what you need.

**Comment by id** [comment]
`{badmatch,9638}` confirms that this is exactly the same issue.
Not rude at all, we plan to release it at the end of this week or early next week.

**Comment by lotros** [comment]
> `{badmatch,9638}` confirms that this is exactly the same issue. Not rude at all, we plan to release it at the end of this week or early next week.

Great, I am glad its apart of that problem, because we were concerned that the fix you've suggested in other similar issues to limit the $SYS topic didn't achieve any result and we though it might be another issue. The timeframe is great, we've made an update to our application that lowered the reconnection time to compensate for the loss of connection, so we can manage the issue until the release of 5.8.1.

**Comment by lotros** [comment]
Update to 5.8.1 definitely solved our issue. Thank you a lot for the quick response and keeping the timeframe as you said.
